pkg/engine/factories/contextloaderfactory.go

```diff
@@ -8,79 +8,131 @@ import (
 	kyvernov1 "github.com/kyverno/kyverno/api/kyverno/v1"
 	engineapi "github.com/kyverno/kyverno/pkg/engine/api"
 	enginecontext "github.com/kyverno/kyverno/pkg/engine/context"
+	"github.com/kyverno/kyverno/pkg/engine/context/loaders"
 	"github.com/kyverno/kyverno/pkg/engine/jmespath"
 	"github.com/kyverno/kyverno/pkg/logging"
 )
 
-func DefaultContextLoaderFactory(cmResolver engineapi.ConfigmapResolver) engineapi.ContextLoaderFactory {
-	return func(policy kyvernov1.PolicyInterface, rule kyvernov1.Rule) engineapi.ContextLoader {
-		return &contextLoader{
-			logger:     logging.WithName("DefaultContextLoaderFactory"),
-			cmResolver: cmResolver,
+type ContextLoaderFactoryOptions func(*contextLoader)
+
+func DefaultContextLoaderFactory(opts ...ContextLoaderFactoryOptions) engineapi.ContextLoaderFactory {
+	return func(policy, rule string) engineapi.ContextLoader {
+		cl := &contextLoader{
+			logger:       logging.WithName("DefaultContextLoaderFactory"),
+			initializers: make([]engineapi.Initializer, 0),
+		}
+		for _, o := range opts {
+			o(cl)
+		}
+		if cl.jp == nil {
+			cl.jp = jmespath.NewWithDefaults()
 		}
+		return cl
+	}
+}
+
+// WithAPIClient enables loading of APICall context entries
+func WithAPIClient(client engineapi.RawClient) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.client = client
+	}
+}
+
+// WithRegistryClientFactory enables loading of ImageData context entries
+func WithRegistryClientFactory(rcf engineapi.RegistryClientFactory) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.rclientFactory = rcf
+	}
+}
+
+// WithConfigMapResolver enables loading of ConfigMap context entries
+func WithConfigMapResolver(cmResolver engineapi.ConfigmapResolver) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.cmResolver = cmResolver
+	}
+}
+
+func WithInitializer(initializer engineapi.Initializer) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.initializers = append(cl.initializers, initializer)
+	}
+}
+
+// WithJMESPath sets the JMESPath engine
+func WithJMESPath(jp jmespath.Interface) ContextLoaderFactoryOptions {
+	return func(cl *contextLoader) {
+		cl.jp = jp
 	}
 }
 
 type contextLoader struct {
-	logger     logr.Logger
-	cmResolver engineapi.ConfigmapResolver
+	logger         logr.Logger
+	jp             jmespath.Interface
+	client         engineapi.RawClient
+	rclientFactory engineapi.RegistryClientFactory
+	cmResolver     engineapi.ConfigmapResolver
+	initializers   []engineapi.Initializer
 }
 
 func (l *contextLoader) Load(
 	ctx context.Context,
-	jp jmespath.Interface,
```

Comment:
why changing that ? makes the loader more stateful

Reply:
To me common state like the JMESPath engine, registry clients,  CM resolvers, etc. seem like things that should be initialized once when the factory is created, and reused when a client needs a new instance. None of these items need to change across the request lifecycle. Otherwise why have a factory?

Reply:
Factory was introduced to allow creating different types of context loaders (cli used a custom factory, playground too).

to me it looks like the loader does not own these objects, it only manages the loading part using what we pass to it.

Reply:
I think the question is between adding these when initializing a `DefaultContextLoaderFactory()` or invoking the `load` method on the factory. In either case the loader gets passed the values.

Reply:
I think those objects belong to the engine.

Reply:
Perhaps its the naming, but I am thinking of the factory as a way to drive uniform creation, and simplify creation, and hence felt these static objects should be set once in the factory.

Since its not related to the core changes required, I am OK with removing and discussing in a separate PR.

Reply:
The loader knows how to use a registry client to load image data in the context for example, but it doesn’t mean it owns the registry client.

it’s just responsible for the loading logic. There’s no reason to make them coupled IMO.

Reply:
> I think those objects belong to the engine.

Yes, that may be best.

Reply:
> The loader knows how to use a registry client to load image data in the context for example, but it doesn’t mean it owns the registry client.
> 
> it’s just responsible for the loading logic. There’s no reason to make them coupled IMO.

Its using a registry client factory.

Reply:
It was an example … but it’s the same thing, the registry client factory belongs to the engine, like the jp interface, client interface, etc…

The only interface specific to the context loader is the configmap resolver (it is only consumed by the context loader).